control/controlclient/direct.go

```diff
@@ -1204,6 +1218,67 @@ func sleepAsRequested(ctx context.Context, logf logger.Logf, timeoutReset chan<-
 	}
 }
 
+// getNoiseClient returns the noise client, creating one if one doesn't exist.
+func (c *Direct) getNoiseClient(ctx context.Context) (*noiseClient, error) {
+	c.mu.Lock()
+	su := c.serverURL
+	np := c.noiseClient
+	sk := c.serverNoiseKey
+	c.mu.Unlock()
+	if sk.IsZero() {
+		return nil, errors.New("zero serverNoiseKey")
+	}
+	if np != nil {
+		return np, nil
+	}
+	k, err := c.getMachinePrivKey()
+	if err != nil {
+		return nil, err
+	}
+
+	np, err = newNoiseClient(ctx, k, sk, su)
+	if err != nil {
+		return nil, err
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.noiseClient != nil {
+		// We lost the race, clean up.
+		np.Close()
+		np = c.noiseClient
+	} else {
+		c.noiseClient = np
+	}
+	return np, nil
+}
+
+func (c *Direct) setDNSNoise(ctx context.Context, req *tailcfg.SetDNSRequest) error {
```

Comment:
req.Version = tailcfg.CurrentCapabilityVersion?

just document that this mutates req?

Reply:
created a copy of req